CMakeLists.txt

```diff
@@ -72,37 +75,37 @@ set(GCC_VERSION 11.2.0)
 set(LLVM_VERSION 12.0.1)
 
 add_custom_target(qemu-image
-    COMMAND ${CMAKE_COMMAND} -E env "SERENITY_SOURCE_DIR=${CMAKE_SOURCE_DIR}" "SERENITY_ARCH=${SERENITY_ARCH}" "USE_CLANG_TOOLCHAIN=$<BOOL:${USE_CLANG_TOOLCHAIN}>" "LLVM_VERSION=${LLVM_VERSION}" ${CMAKE_SOURCE_DIR}/Meta/build-image-qemu.sh
+    COMMAND ${CMAKE_COMMAND} -E env "SERENITY_SOURCE_DIR=${SERENITY_PROJECT_ROOT}" "SERENITY_ARCH=${SERENITY_ARCH}" "USE_CLANG_TOOLCHAIN=$<BOOL:${USE_CLANG_TOOLCHAIN}>" "LLVM_VERSION=${LLVM_VERSION}" ${SERENITY_PROJECT_ROOT}/Meta/build-image-qemu.sh
     BYPRODUCTS ${CMAKE_BINARY_DIR}/_disk_image
     USES_TERMINAL
 )
 add_custom_target(grub-image
-    COMMAND ${CMAKE_COMMAND} -E env "SERENITY_SOURCE_DIR=${CMAKE_SOURCE_DIR}" "SERENITY_ARCH=${SERENITY_ARCH}" "USE_CLANG_TOOLCHAIN=$<BOOL:${USE_CLANG_TOOLCHAIN}>" "LLVM_VERSION=${LLVM_VERSION}" ${CMAKE_SOURCE_DIR}/Meta/build-image-grub.sh
+    COMMAND ${CMAKE_COMMAND} -E env "SERENITY_SOURCE_DIR=${SERENITY_PROJECT_ROOT}" "SERENITY_ARCH=${SERENITY_ARCH}" "USE_CLANG_TOOLCHAIN=$<BOOL:${USE_CLANG_TOOLCHAIN}>" "LLVM_VERSION=${LLVM_VERSION}" ${SERENITY_PROJECT_ROOT}/Meta/build-image-grub.sh
     BYPRODUCTS ${CMAKE_BINARY_DIR}/grub_disk_image
     USES_TERMINAL
 )
 add_custom_target(extlinux-image
-    COMMAND ${CMAKE_COMMAND} -E env "SERENITY_SOURCE_DIR=${CMAKE_SOURCE_DIR}" "SERENITY_ARCH=${SERENITY_ARCH}" "USE_CLANG_TOOLCHAIN=$<BOOL:${USE_CLANG_TOOLCHAIN}>" "LLVM_VERSION=${LLVM_VERSION}" ${CMAKE_SOURCE_DIR}/Meta/build-image-extlinux.sh
+    COMMAND ${CMAKE_COMMAND} -E env "SERENITY_SOURCE_DIR=${SERENITY_PROJECT_ROOT}" "SERENITY_ARCH=${SERENITY_ARCH}" "USE_CLANG_TOOLCHAIN=$<BOOL:${USE_CLANG_TOOLCHAIN}>" "LLVM_VERSION=${LLVM_VERSION}" ${SERENITY_PROJECT_ROOT}/Meta/build-image-extlinux.sh
     BYPRODUCTS ${CMAKE_BINARY_DIR}/extlinux_disk_image
     USES_TERMINAL
 )
 
 add_custom_target(lint-shell-scripts
-    COMMAND ${CMAKE_SOURCE_DIR}/Meta/lint-shell-scripts.sh
+    COMMAND ${SERENITY_PROJECT_ROOT}/Meta/lint-shell-scripts.sh
     USES_TERMINAL
 )
 add_custom_target(check-style
-    COMMAND ${CMAKE_SOURCE_DIR}/Meta/check-style.sh
+    COMMAND ${SERENITY_PROJECT_ROOT}/Meta/check-style.sh
     USES_TERMINAL
 )
 
 add_custom_target(install-ports
-    COMMAND ${CMAKE_COMMAND} -E env "SERENITY_SOURCE_DIR=${CMAKE_SOURCE_DIR}" "SERENITY_ARCH=${SERENITY_ARCH}" ${CMAKE_SOURCE_DIR}/Meta/install-ports-tree.sh
+    COMMAND ${CMAKE_COMMAND} -E env "SERENITY_SOURCE_DIR=${SERENITY_PROJECT_ROOT}" "SERENITY_ARCH=${SERENITY_ARCH}" ${SERENITY_PROJECT_ROOT}/Meta/install-ports-tree.sh
     USES_TERMINAL
 )
 
 add_custom_target(configure-components
-    COMMAND ConfigureComponents
+    COMMAND $<TARGET_FILE:ConfigureComponents>
     DEPENDS ConfigureComponents
```

Comment:
Aha makes sense. I removed a majority of the bogus DEPENDS, but it's possible I misunderstood the rules about files in the source tree, e.g. some of the shell scripts we use in these custom command/custom target commands.

Reply:
Only [add_custom_command](https://cmake.org/cmake/help/latest/command/add_custom_command.html) can establish target-level dependency with its `DEPENDS` argument according to the documentation. You should also think about what is a concern during configure time and build time. For example, I have the generator script that's being invoked in this [example](https://github.com/friendlyanon/cmake-init-gif-engine/blob/master/CMakeLists.txt#L29-L30) listed as a dependency, such that when that script is updated, I get proper dependency tracking. Something like this was also missing for that write_if_different script that was in use before.

Reply:
Hmm so, it's incorrect to remove DEPENDS IPCCompiler from this custom command, because the target-level dependency added by having `$<TARGET_FILE:IPCCompiler>`  in a COMMAND statement doesn't cause the command to be re-run when we re-compile IPCCompiler?

```
    add_custom_command(
        OUTPUT ${output}
        BYPRODUCTS ${output}.tmp
        COMMAND $<TARGET_FILE:IPCCompiler> ${source} > ${output}.tmp
        COMMAND "${CMAKE_COMMAND}" -E copy_if_different ${output}.tmp ${output}
        COMMAND "${CMAKE_COMMAND}" -E remove ${output}.tmp
-       DEPENDS IPCCompiler
        VERBATIM
        MAIN_DEPENDENCY ${source}
    )
```

> This target-level dependency does NOT add a file-level dependency that would cause the custom command to re-run whenever the executable is recompiled. List target names with the DEPENDS option to add such file-level dependencies.

And for the add_custom_target, we should `add_dependencies(configure-components ConfigureComponents)` ?

Reply:
I saw you put a FIXME for `ZIP_LISTS`, it might be worth putting one here as well, since `$<TARGET_FILE:...>` does not establish a target-level dependency if [CMP0112](https://cmake.org/cmake/help/latest/policy/CMP0112.html) is set to `NEW` (i.e. when a version >=3.19 is requested). I don't know if Serenity will ever have a reason to do that, but it might be worth keeping in mind, even if you consider incorrect parallel builds for ExternalProject before 3.19, for which there is an [easy fix](https://github.com/friendlyanon/cmake-init-gif-engine/commit/662ca8894d833c8c945c537c5d97d95dc6ffa4cf).

Reply:
When I read the documentation on those target-level dependencies I read it as, "TARGET_FILE, TARGET_LINKER_FILE, TARGET_SONAME_FILE, and TARGET_PDB_FILE establish a target level dependency, but with CMP0112 set to NEW, these extended generator expressions with _NAME, _DIR, and _BASE_NAME will stop adding the target dependency". I take it this means that TARGET_FILE actually is some combo of TARGET_FILE_NAME and TARGET_FILE_DIR?

There's a large chunk of refactoring left to get the ExternalProject stuff up and going to have proper CMakeToolchain support for the target build.

Regarding upgrading versions, I asked around in our discord and the consensus was that until there's some killer feature that we can't do without, it's preferred to stay on a CMake version supported by the current major LTS releases (Ubuntu 20.04 seems to be sticking with 3.16 through the ubuntu PPAs).

> I guess this already sort of answers your question.

Also yeah, pretty much. I'll be adding the dependencies manually in a push soon ™️ .